layers/+window-management/exwm/funcs.el

```diff
@@ -78,6 +78,59 @@ Can show completions at point for COMMAND using helm"
   (interactive)
   (start-process "" nil exwm-locking-command))
 
+(defun exwm/autostart-process (name command &optional directory)
+  "Start a program is subprocess and append it in `exwm-autostart-process-list'.
+NAME is name for process.
+COMMAND is the shell command to run.
+DIRECTORY is the working directory in which the process is run.  It defaults
+  `user-home-directory' if not provided.
+The started process is also added to `exwm-autostart-process-list'."
+  (add-to-list (let ((default-directory (or directory user-home-directory)))
+                 (start-process-shell-command name nil command))
+               exwm-autostart-process-list))
+
+(defun exwm//start-desktop-application (xdg)
+  "Autostart an application from a XDG desktop entry specification."
+  (when-let* ((type (gethash "Type" xdg))
+              (_application? (string= type "Application"))
+              ;; Note that we can't always assume the presence of Exec: when
+              ;; DBusActivatable support is added, the application is launched
+              ;; through D-Bus rather than a program.
+              (basename (gethash "Name" xdg))
+              (cmd (gethash "Exec" xdg))
+                            ;; (dbus-p (gethash "DBusActivatable" xdg)) ; TODO: support
+              (directory (gethash "Path" xdg) user-home-directory)
+              (_include? (and (null (gethash "Hidden" xdg))
+                              (if-let ((only-show (gethash "OnlyShowIn" xdg)))
+                                  (member "EXWM" (split-string only-show ";" t))
+                                t)
+                              (if-let ((not-show (gethash "NotShowIn" xdg)))
+                                  (not (member "EXWM" (split-string not-show ";" t)))
+                                t)
+                              (if-let ((try-exec (gethash "TryExec" xdg)))
+                                  (executable-find try-exec)
+                                t))))
+    (exwm/autostart-process basename cmd directory)))
+
+(defun exwm//read-xdg-autostart-files ()
+  "Return a list for autostart applications as defined by the freedesktop specification."
+  (cl-loop with xdg-specs = (make-hash-table :test 'equal)
+           for dir in (append (xdg-config-dirs) (list (xdg-config-home)))
+           for autostart-dir = (concat dir "/autostart")
+           for autostart-files = (when (file-exists-p autostart-dir)
+                                   (directory-files autostart-dir t ".+\\.desktop$"))
+           do
+           (cl-loop for file in autostart-files do
+                    (push (xdg-desktop-read-file file) xdg-specs))
+           finally (return xdg-specs)))
+
+(defun exwm//autostart-xdg-applications ()
+  "Run the autostart applications as defined by the freedesktop autostart specification."
+  (unless exwm-autostart-process-list
+    (cl-loop for xdg in (exwm//read-xdg-autostart-files)
```

Comment:
Related to last comment.
Here you appear to use the return value of `exwm//read-xdg-autostart-files` as a list of hash table.

---

In addition, you don't need to use `cl-loop` any more
```elisp
(mapc #'exwm//start-desktop-application (exwm//read-xdg-autostart-files))
```